cdap-security/src/test/java/co/cask/cdap/security/authorization/InMemoryAuthorizer.java

```diff
@@ -172,20 +199,38 @@ public void removeRoleFromPrincipal(Role role, Principal principal) throws RoleN
   }
 
   private Set<Privilege> getPrivileges(Principal principal) {
-    Set<Privilege> privileges = new HashSet<>();
-    Set<Map.Entry<EntityId, Set<Action>>> entries = table.column(principal).entrySet();
-    for (Map.Entry<EntityId, Set<Action>> entry : entries) {
-      for (Action action : entry.getValue()) {
-        privileges.add(new Privilege(entry.getKey(), action));
+    readLock.lock();
+    try {
+      Set<Privilege> privileges = new HashSet<>();
+      Set<Map.Entry<EntityId, Set<Action>>> entries = table.column(principal).entrySet();
+      for (Map.Entry<EntityId, Set<Action>> entry : entries) {
+        for (Action action : entry.getValue()) {
+          privileges.add(new Privilege(entry.getKey(), action));
+        }
       }
+      return privileges;
+    } finally {
+      readLock.unlock();
     }
-    return privileges;
   }
 
   private Set<Action> getActions(EntityId entity, Principal principal) {
-    if (!table.contains(entity, principal)) {
-      table.put(entity, principal, new HashSet<Action>());
+    Set<Action> actions;
+    readLock.lock();
+    try {
+      actions = table.get(entity, principal);
+    } finally {
+      readLock.unlock();
+    }
+    if (actions == null) {
```

Comment:
This can be a race since you released the read lock, someone can acquire the write lock and updated the table. Then in the following block, you'll overwrite that.